c_src/erl_optics.c

```diff
@@ -46,14 +48,49 @@ static int load(ErlNifEnv *env, void **priv_data, ERL_NIF_TERM load_info) {
     atom_dist = enif_make_atom(env, "dist");
     atom_histo = enif_make_atom(env, "histo");
     atom_quantile = enif_make_atom(env, "quantile");
-    poller = optics_poller_alloc();
-    ctx = enif_alloc(sizeof(struct map_ctx));
-    if (poller) {
-      optics_poller_backend(poller, (void *) ctx, backend_eo, backend_free_eo);
-    }
+    carbon_poller = 0;
+    erlang_poller = 0;
     return 0;
 }
 
+
+static ERL_NIF_TERM eo_allocate_carbon_poller(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+    if (!poller){
+        poller = optics_poller_alloc();
+    }
+    if (!poller) return ERROR("poller_allocation_error");
+    if (carbon_poller == 0) {
+        size_t host_size = 255;
+        size_t port_size = 255;
+        char host[host_size];
+        char port[port_size];
+        if (enif_get_string(env, argv[1], host, host_size, ERL_NIF_LATIN1) < 1) return ERROR("invalid_hostname");
+        if (enif_get_string(env, argv[2], port, port_size, ERL_NIF_LATIN1) < 1) return ERROR("invalid_port");
+        optics_dump_carbon(poller, host, port);
+        carbon_poller = 1;
+    }
+    return atom_ok;
+}
+
+static ERL_NIF_TERM eo_allocate_erlang_poller(
+    ErlNifEnv *env, int argc, const ERL_NIF_TERM argv[])
+{
+    if (!poller){
+        poller = optics_poller_alloc();
```

Comment:
I would prefer if we returned an error if the poller is already allocated. The reasoning being that the purpose of this function is to allocate the poller which means that if it's already allocated then I would consider that a user error. I generally think it's a better idea to be strict then permissive in interfaces as it tends to surface errors more quickly.

Reply:
So re-reading a bit, I think I understand why it's structured this way as you can either call `erlang_poller` or `carbon_poller` or both. In which case the interface works but I would then argue as to whether there's ever a need to have both.